rom/laghos_rom.cpp

```diff
@@ -597,9 +695,274 @@ void ROM_Sampler::SetupEQP_Force_Eq(const CAROM::Matrix* snapX,
                       std::to_string(rank));
 }
 
-void ROM_Sampler::SetupEQP_Force(const CAROM::Matrix* snapX, const CAROM::Matrix* snapV, const CAROM::Matrix* snapE,
-                                 const CAROM::Matrix* basisV, const CAROM::Matrix* basisE,
-                                 ROM_Options const& input)
+// Compute the reduced quadrature rule for the energy-conserving EQP case.
+void ROM_Sampler::SetupEQP_En_Force_Eq(const CAROM::Matrix* snapX,
+									   const CAROM::Matrix* snapV,
+									   const CAROM::Matrix* snapE,
+									   const CAROM::Matrix* basisV,
+                                       const CAROM::Matrix* basisE,
+                                       ROM_Options const& input)
+{
+	const IntegrationRule *ir0 = input.FOMoper->GetIntegrationRule();
+	const int nqe = ir0->GetNPoints();
+	const int ne = input.H1FESpace->GetNE();
+	const int NQ = ne * nqe;
+
+	const int NBv = basisV->numColumns();
+	const int NBe = basisE->numColumns();
+	const int NBmin = min(NBv, NBe);
+
+	Array<int> numSnapVar(3);
+	numSnapVar[0] = snapX->numColumns();
+	numSnapVar[1] = snapV->numColumns();
+	numSnapVar[2] = snapE->numColumns();
+
+	const int nsnap = numSnapVar.Max();
+
+	Array<int> numSkipped(3);
+	for (int i=0; i<3; ++i) numSkipped[i] = nsnap - numSnapVar[i];
+	MFEM_VERIFY(numSkipped.Max() <= 1, "");
+
+	Vector rv(nqe), re(nqe);
+
+	// G is the matrix of accuracy constraints used to enforce that the
+	// evaluated quantity remains close to the result of the full quadrature rule.
+
+	// Declare G of size ((NBv + NBe) * (nsnap+1)) x NQ; store its transpose Gt.
+	// The first NBv*(nsnap+1) rows of G hold the velocity constraints;
+	// the remaining NBe*(nsnap+1) rows hold the energy constraints.
+	CAROM::Matrix Gt(NQ, (NBv + NBe) * (nsnap+1), true);
+
+	// row index of G where energy constraints start
+	const int estart = NBv * (nsnap + 1);
+
+	Vector v_i(tH1size), x_i(tH1size), e_i(tL2size);
+	Vector w_j_e, v_i_e, v_j_e;
+
+	Vector S((2*input.H1FESpace->GetVSize()) + input.L2FESpace->GetVSize());
+	Vector S_v(S, input.H1FESpace->GetVSize(), input.H1FESpace->GetVSize());  // Subvector
+
+	MFEM_VERIFY(tH1size == basisV->numRows(), "");
+	MFEM_VERIFY(tL2size == basisE->numRows(), "");
+	CAROM::Matrix Wv(H1size, NBv, true);
+	CAROM::Matrix We(L2size, NBe, true);
+
+	// jth Wv columnn holds the jth velocity basis vector
+	// TODO: do we need Wv anymore? it's not used anywhere
+	for (int j=0; j<NBv; ++j)
+	{
+		for (int i=0; i<tH1size; ++i) v_i[i] = (*basisV)(i,j);
+		gfH1.SetFromTrueDofs(v_i);
+		for (int i=0; i<H1size; ++i) Wv(i,j) = gfH1[i];
+	}
```

Comment:
Given the changes made in our choice for force vectors, I believe that `Wv` is not used anymore. It was used in the previous EQP implementation to form the velocity force vector.